traits/trait_types.py

```diff
@@ -2283,16 +2283,41 @@ def __init__(self, *types, **metadata):
             if len(types) == 0:
                 types = [Trait(element) for element in default_value]
 
-        self.types = tuple([trait_from(type) for type in types])
+        self.types = tuple(trait_from(type) for type in types)
         self.init_fast_validate(ValidateTrait.tuple, self.types)
 
         if default_value is None:
-            default_value = tuple(
-                [type.default_value()[1] for type in self.types]
+            # Optimisation: if all child traits have a constant default value,
+            # we can use a constant default value too. Otherwise the default
+            # needs to be computed dynamically.
+            child_defaults = []
+            child_default_types = []
+            for child_trait in self.types:
+                child_default_type, child_default = child_trait.default_value()
+
+                child_default_types.append(child_default_type)
+                child_defaults.append(child_default)
+
+            constant_default = all(
+                dvt == DefaultValue.constant for dvt in child_default_types
             )
```

Comment:
Short-circuit the loop if you hit a non-constant default
```suggestion
            child_defaults = []
            constant_default = True
            for child_trait in self.types:
                child_default_type, child_default = child_trait.default_value()
                if child_default_type != DefaultValue.constant:
                    constant_default = False
                    break
                child_defaults.append(child_default)
```
or something like this?

Reply:
I actually already did this, and then changed to the current implementation on the basis that it was clearer and that the loop seemed unlikely to be a bottleneck.

Reply:
(Off-topic, but turning a list of pairs into a pair of lists seems unnecessarily awkward in Python - it feels as though it ought to be a one-liner.)